### ISSUE TITLE
Development and Release Branches

### ISSUE THREAD

**Issue by FabrizioMoggio** [camaraproject/Commonalities#66]
**Problem description**
in the Camara Versioning Guidelines document (https://github.com/camaraproject/Commonalities/blob/main/documentation/Camara_Versioning_Guidelines.md) it is described how to create Release Branches. It is not clear to me where the development is done (Release Branches or Main). I see two options and I'm not sure which is the one adopted by CAMARA.

Option 1:

- the **Main Branch** is where the External Developer can find the very last **released version** of a Family and, using the Release Tags, also the previous releases. 
- the **Release Branches** are meant for **development** of a new release. When the development is closed the code is merged in Main. Considering an API Family with more APIs inside, when all the APIs are merged into Main a new Release Tag, in Main, is added.

pro: 
-it is possible to develop in parallel different releases
-it is easier for the External Developer to find the very last version (Main Branch)

Option 2:

- the **Main Branch** is where the External Developer can find the work in progress. It is meant **for development**. When the work is consolidated a Release Tag is created and a Release Branch is created. Considering an API Family with more APIs inside, a Release Tag and a new Release Branch is done when all the APIs are completed.
- the Release Branches is where the External Developer can find a stable version. 

**Expected action**
Clarify this in the  Camara Versioning Guidelines document

**Comment by jlurien** [comment]
Hi @FabrizioMoggio. I think that option 2 is the approach we are following, at least in the WGs where I'm most directly involved, QoD and DeviceLocation. The API specs in main branch have the -wip suffix added to the version number, which is the next version under development. When stable and approved, release branch and tag are created, removing the -wip suffix. From that moment, next change increases version number in main with the -wip added.

**Comment by rartych** [comment]
Yes, option 2, current [Versioning_Guidelines](https://github.com/camaraproject/Commonalities/blob/main/documentation/Camara_Versioning_Guidelines.md)  state that:

> Main branch is assumed to be the latest.

I have added "release-management" label to this issues, as there are a number of issues with this label indicating the need to explain the whole process. Contributions are welcome!

**Comment by FabrizioMoggio** [comment]
the problem I see is how to work on two versions in parallel. I can see this is not a common, or anyway a very good, practice :-), but It can be needed. For example for the TI API we are consolidating version 1 while working on version 2.

**Comment by FabrizioMoggio** [comment]
Another comment is:  why we need the branches for releases if we already have Release Tags in the Main branch? The original idea (I was the one proposing it) was to have Release Branches for internal development and Main for external user where they can find all the releases using Release Tags. Main, in my original proposal, was the "safe place" where the External Developer can find the last, and previous, releases

**Comment by hdamker** [comment]
Hi @FabrizioMoggio ... first to be confirm that we follow option two. It's common practice in open source project that pull requests are going against main. Using development branches is a concept used in organization internal projects. 

> Another comment is: why we need the branches for releases if we already have Release Tags in the Main branch?

The release branch has two functions:
* used within links to the files in a release, like e.g. in https://github.com/camaraproject/QualityOnDemand/blob/release-0.9.0/code/API_definitions/qod-api.yaml
* branch point in case there will be patch releases, e.g a (hypothetical) v0.9.1 with documentation corrections

Please be aware that the release version and branch is always for the complete repository. If you have two different major versions of an API (which are essentially then different APIs) you have to develop them in different files and with different base paths, e.g. apixy/v1 and apixy/v2alpha or similar.

Regards
Herbert

**Comment by FabrizioMoggio** [comment]
In my understanding, this Issue has the following **conclusion**:

- **the Main Branch** is where the External Developer can find the **work in progress**. It is meant for development. When the work is consolidated a Release Tag is created and a Release Branch is created. Considering an API Family with more APIs inside, a Release Tag and a new Release Branch is done when all the APIs are completed.

- the **Release Branches** is where the External Developer can find a **stable version**.

The Issue will be closed when this will be clarified in the Camara Versioning Guidelines document
